passes/cmds/chformal.cc

```diff
@@ -278,21 +352,68 @@ struct ChformalPass : public Pass {
 			if (mode =='p')
 			{
 				for (auto cell : constr_cells)
-					module->addCover(NEW_ID_SUFFIX("coverenable"),
-						cell->getPort(ID::EN), State::S1, cell->get_src_attribute());
+				{
+					if (cell->type == ID($check)) {
+						Cell *cover = module->addCell(NEW_ID_SUFFIX("coverenable"), ID($check));
+						cover->attributes = cell->attributes;
+						cover->parameters = cell->parameters;
+						for (auto conn : cell->connections())
+							cover->setPort(conn.first, conn.first == ID::A ? State::S1 : conn.second);
+					} else {
+						module->addCover(NEW_ID_SUFFIX("coverenable"),
+							cell->getPort(ID::EN), State::S1, cell->get_src_attribute());
+					}
+				}
 			}
 			else
 			if (mode == 'c')
 			{
-				for (auto cell : constr_cells)
-					if (assert2assume && cell->type == ID($assert))
-						cell->type = ID($assume);
-					else if (assume2assert && cell->type == ID($assume))
-						cell->type = ID($assert);
-					else if (live2fair && cell->type == ID($live))
-						cell->type = ID($fair);
-					else if (fair2live && cell->type == ID($fair))
-						cell->type = ID($live);
+				for (auto cell : constr_cells) {
+					IdString flavor = formal_flavor(cell);
+					if (assert2assume && flavor == ID($assert))
+						set_formal_flavor(cell, ID($assume));
+					else if (assume2assert && flavor == ID($assume))
+						set_formal_flavor(cell, ID($assert));
+					else if (live2fair && flavor == ID($live))
+						set_formal_flavor(cell, ID($fair));
+					else if (fair2live && flavor == ID($fair))
+						set_formal_flavor(cell, ID($live));
+				}
+			}
+			else
+			if (mode == 'l')
+			{
+				for (auto cell : constr_cells) {
+					if (cell->type != ID($check))
+						continue;
+
+					if (is_triggered_check_cell(cell))
+						log_error("Cannot lower edge triggered $check cell %s, run async2sync or clk2fflogic first.\n", log_id(cell));
+
+
+					Cell *plain_cell = module->addCell(NEW_ID, formal_flavor(cell));
+
+					plain_cell->attributes = cell->attributes;
+
+					SigBit id_a = cell->getPort(ID::A);
+					SigBit id_en = cell->getPort(ID::EN);
+
+					plain_cell->setPort(ID::A, id_a);
+					plain_cell->setPort(ID::EN, id_en);
+
+					SigBit combined_en = module->And(NEW_ID, id_a, id_en);
```

Comment:
Yeah `A` here should be inverted for assert and assume. While `live` and `fair` are a form of assert and assume, they can only be violated for infinite traces where `A` remains low forever, so printing a message whenever there is progress in keeping the assertion/assumption valid seems like the only useful thing.

Reply:
For `live` and `fair` you would want to print the fact that they were active every time they were active so I think it makes sense to also print the message in that instance as it might contain relevant signals as arguments.

For `cover` when using SV syntax to manually print a message you would also get that message every time it is reached.

I think in some cases it makes sense to print only the first time for cover, but I'd want that to be user selectable and I think the mode where it prints every time a cover is reached should be the default.

Reply:
This is during explicit conversion from `$check` cells to `$assert` etc using `chformal -lower` where I currently do add a `$print` cell when the `$check` cell has a message. This is unrelated to the verilog frontend or backend and only about not losing messages that may be present on `$check` cells during this conversion.